File: services/platform-api/cmd/server/billing.go
package main

import (
	"encoding/json"
	"net/http"

	"github.com/Techshrr/GoJet_Short_Link/app/billing"
)

func (s *server) workspaceBilling(w http.ResponseWriter, r *http.Request) {
	wid, err := pathID(r, "id")
	if err != nil {
		jsonResponse(w, 400, map[string]string{"error": "工作区编号无效"})
		return
	}
	if _, err = s.workspace.Role(r.Context(), wid, currentUser(r).ID); err != nil {
		jsonResponse(w, 403, map[string]string{"error": "无权查看该工作区账单"})
		return
	}
	plans, err := s.billing.Plans(r.Context(), false)
	if err != nil {
		jsonResponse(w, 503, map[string]string{"error": "套餐暂时不可用"})
		return
	}
	sub, invoices, err := s.billing.Subscription(r.Context(), wid)
	if err != nil {
		jsonResponse(w, 503, map[string]string{"error": "订阅信息暂时不可用"})
		return
	}
	jsonResponse(w, 200, map[string]any{"plans": plans, "subscription": sub, "invoices": invoices})
}

func (s *server) requestInvoice(w http.ResponseWriter, r *http.Request) {
	wid, err := pathID(r, "id")
	var in struct {
		PlanCode string `json:"plan_code"`
		Type     string `json:"type"`
	}
	if decode(w, r, &in) != nil {
		return
	}
	if err != nil {
		jsonResponse(w, 400, map[string]string{"error": "工作区编号无效"})
		return
	}
	invoice, err := s.billing.Request(r.Context(), currentUser(r).ID, wid, in.PlanCode, in.Type)
	if err != nil {
		jsonResponse(w, 422, map[string]string{"error": err.Error()})
		return
	}
	jsonResponse(w, 201, invoice)
}

func (s *server) cancelSubscription(w http.ResponseWriter, r *http.Request) {
	wid, err := pathID(r, "id")
	var in struct {
		Cancel bool `json:"cancel"`
	}
	if decode(w, r, &in) != nil {
		return
	}
	if err != nil {
		jsonResponse(w, 400, map[string]string{"error": "工作区编号无效"})
		return
	}
	if err = s.billing.CancelAtPeriodEnd(r.Context(), currentUser(r).ID, wid, in.Cancel); err != nil {
		jsonResponse(w, 422, map[string]string{"error": err.Error()})
		return
	}
	jsonResponse(w, 200, map[string]bool{"cancel_at_period_end": in.Cancel})
}

func (s *server) adminPlans(w http.ResponseWriter, r *http.Request) {
	plans, err := s.billing.Plans(r.Context(), true)
	if err != nil {
		jsonResponse(w, 503, map[string]string{"error": "套餐列表暂时不可用"})
		return
	}
	jsonResponse(w, 200, map[string]any{"data": plans})
}

func (s *server) adminUpdatePlan(w http.ResponseWriter, r *http.Request) {
	id, err := pathID(r, "id")
	var raw struct {
		Name                   string          `json:"name"`
		Description            string          `json:"description"`
		Status                 string          `json:"status"`
		MonthlyPriceCents      int64           `json:"monthly_price_cents"`
		LinkLimit              int64           `json:"link_limit"`
		QRLimit                int64           `json:"qr_limit"`
		TextLimit              int64           `json:"text_limit"`
		BioLimit               int64           `json:"bio_limit"`
		FileStorageBytes       int64           `json:"file_storage_bytes"`
		MemberLimit            int64           `json:"member_limit"`
		AnalyticsRetentionDays int             `json:"analytics_retention_days"`
		Features               json.RawMessage `json:"features"`
	}
	if decode(w, r, &raw) != nil {
		return
	}
	if err != nil {
		jsonResponse(w, 400, map[string]string{"error": "套餐编号无效"})
		return
	}
	in := billing.PlanInput{Name: raw.Name, Description: raw.Description, Status: raw.Status, MonthlyPriceCents: raw.MonthlyPriceCents, LinkLimit: raw.LinkLimit, QRLimit: raw.QRLimit, TextLimit: raw.TextLimit, BioLimit: raw.BioLimit, FileStorageBytes: raw.FileStorageBytes, MemberLimit: raw.MemberLimit, AnalyticsRetentionDays: raw.AnalyticsRetentionDays, Features: raw.Features}
	if err = s.billing.UpdatePlan(r.Context(), id, in); err != nil {
		jsonResponse(w, 422, map[string]string{"error": err.Error()})
		return
	}
	jsonResponse(w, 200, map[string]bool{"updated": true})
}

func (s *server) adminInvoices(w http.ResponseWriter, r *http.Request) {
	limit, offset := page(r)
	items, err := s.billing.Invoices(r.Context(), r.URL.Query().Get("status"), limit, offset)
	if err != nil {
		jsonResponse(w, 503, map[string]string{"error": "账单列表暂时不可用"})
		return
	}
	jsonResponse(w, 200, map[string]any{"data": items})
}

func (s *server) adminSettleInvoice(w http.ResponseWriter, r *http.Request) {
	id, err := pathID(r, "id")
	var in struct {
		Status string `json:"status"`
		Note   string `json:"note"`
	}
	if decode(w, r, &in) != nil {
		return
	}
	if err != nil {
		jsonResponse(w, 400, map[string]string{"error": "账单编号无效"})
		return
	}
	if err = s.billing.Settle(r.Context(), currentAdmin(r).ID, id, in.Status, in.Note); err != nil {
		jsonResponse(w, 422, map[string]string{"error": err.Error()})
		return
	}
	jsonResponse(w, 200, map[string]bool{"updated": true})
}

File: frontend/packages/api-client/src/billing.ts
import type { ApiTransport } from "./links";

export interface BillingPlan {
  id: number; code: string; name: string; monthly_price_cents: number; currency: string; description: string; status: "active" | "archived";
  features: unknown; link_limit: number; qr_limit: number; text_limit: number; bio_limit: number; file_storage_bytes: number; member_limit: number; analytics_retention_days: number;
}
export interface BillingSubscription { workspace_id: number; plan_id: number; plan_code: string; plan_name: string; status: string; period_started_at: string; period_ends_at?: string; cancel_at_period_end: boolean; }
export interface BillingUsage { plan_name: string; plan_code: string; links: number; link_limit: number; qr_codes: number; qr_limit: number; text_shares: number; text_limit: number; bio_pages: number; bio_limit: number; file_bytes: number; file_storage_bytes: number; members: number; member_limit: number; analytics_retention_days: number; }
export interface BillingInvoice { id: number; workspace_id: number; plan_id: number; requested_by: number; source_amount_cents: number; amount_cents: number; fx_markup_bps: number; period_months: number; period_days: number; invoice_number: string; plan_name: string; plan_code: string; invoice_type: string; billing_cycle: string; source_currency: string; currency: string; fx_rate: string; fx_provider: string; fx_quoted_at?: string; status: "pending" | "overdue" | "paid" | "void" | string; due_at: string; paid_at?: string; admin_note?: string; created_at: string; }
export interface WorkspaceBillingPayload { plans: BillingPlan[]; subscription: BillingSubscription; usage: BillingUsage; invoices: BillingInvoice[]; }
export interface PaymentMethod { code: string; name: string; enabled: boolean; mode: string; }
export interface PaymentCheckout { transaction_id: number; provider: string; provider_name: string; mode: string; redirect_url?: string; qr_content?: string; merchant_order_no: string; reused?: boolean; }
export interface AdminPayment { id: number; invoice_id: number; workspace_id: number; owner_email: string; provider: string; merchant_order_no: string; provider_order_id: string; amount_cents: number; currency: string; status: string; failure_reason?: string; paid_at?: string; created_at: string; updated_at: string; }
export interface PaymentCallback { id: number; provider: string; request_id: string; merchant_order_no: string; provider_reference: string; payload_sha256: string; outcome: "accepted" | "rejected"; response_status: number; remote_ip: string; created_at: string; }
export interface AdminPaymentDetail { payment: AdminPayment; callbacks: PaymentCallback[]; payload_redacted: true; }
export interface FXRate { base_currency: string; quote_currency: string; provider: string; rate: string; observed_at: string; expires_at: string; }
export interface FXHistory { invoice_number: string; source_currency: string; currency: string; rate: string; provider: string; markup_bps: number; quoted_at?: string; created_at: string; }
export interface AdminFXPayload { settlement_currency: string; provider: "ecb" | "manual"; markup_bps: number; cache_hours: number; manual_rates: Record<string,string>; rates: FXRate[]; history: FXHistory[]; }
export interface PlanWriteInput { name: string; description: string; status?: "active" | "archived"; monthly_price_cents: number; link_limit: number; qr_limit: number; text_limit: number; bio_limit: number; file_storage_bytes: number; member_limit: number; analytics_retention_days: number; features: unknown; }
export interface PlanCreateInput extends PlanWriteInput { code: string; currency: string; }

export function createBillingClient(api: ApiTransport) {
  return {
    workspace: (workspaceId: number) => api.get<WorkspaceBillingPayload>(`/api/workspaces/${workspaceId}/billing`),
    paymentMethods: (workspaceId: number) => api.get<{data: PaymentMethod[]}>(`/api/workspaces/${workspaceId}/billing/payment-methods`),
    createInvoice: (workspaceId: number, planCode: string, type: "change" | "renew", billingCycle: string) => api.post<BillingInvoice>(`/api/workspaces/${workspaceId}/billing/invoices`, { plan_code: planCode, type, billing_cycle: billingCycle }),
    cancellation: (workspaceId: number, cancel: boolean) => api.patch<{cancel_at_period_end: boolean}>(`/api/workspaces/${workspaceId}/billing/cancellation`, { cancel }),
    checkout: (workspaceId: number, invoiceId: number, provider: string) => api.post<PaymentCheckout>(`/api/workspaces/${workspaceId}/billing/invoices/${invoiceId}/pay`, { provider }),
    adminPlans: () => api.get<{data: BillingPlan[]}>("/api/admin/plans"),
    createPlan: (input: PlanCreateInput) => api.post<{id: number; created: boolean}>("/api/admin/plans", input),
    updatePlan: (planId: number, input: PlanWriteInput) => api.put<{updated: boolean}>(`/api/admin/plans/${planId}`, input),
    archivePlan: (planId: number) => api.post<{archived: boolean}>(`/api/admin/plans/${planId}/archive`),
    adminInvoices: (status = "") => api.get<{data: BillingInvoice[]}>(`/api/admin/invoices${status ? `?status=${encodeURIComponent(status)}` : ""}`),
    settleInvoice: (invoiceId: number, status: "paid" | "void", note: string) => api.post<{updated: boolean}>(`/api/admin/invoices/${invoiceId}/settle`, { status, note }),
    adminPayments: (status = "", provider = "") => api.get<{data: AdminPayment[]}>(`/api/admin/payments?status=${encodeURIComponent(status)}&provider=${encodeURIComponent(provider)}`),
    adminPayment: (paymentId: number) => api.get<AdminPaymentDetail>(`/api/admin/payments/${paymentId}`),
    adminFX: () => api.get<AdminFXPayload>("/api/admin/fx"),
    updateFX: (input: {settlement_currency: string; provider: "ecb" | "manual"; markup_bps: number; cache_hours: number; manual_rates: Record<string,string>; reason: string}) => api.put<{updated: boolean}>("/api/admin/fx", input)
  };
}
